### DISCUSSION TITLE
WinRT.Interop.WindowNative.GetWindowHandle Exception: The Specified Cast is not Valid

### DISCUSSION THREAD

**Discussion by frankayars** [microsoft/microsoft-ui-xaml#7951] [Q&A] [answered]
We are currently encountering an exception with this call:

IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);

The exception message is 'The specified cast is not valid'. The issue is not limited to the parent control. Any UIElement based object passed into that call will cause the exception. Our current project requires the underlying hWnd handle to integrate with 3rd party 3D graphics APIs and any information to resolve this exception would be appreciated.

We have tested different versions of the Windows App SDK (microsoft.windowsappsdk) and get the same result.

**Comment by gautambjain** [comment]
**this** has to be of type Window

Objects of type UIElement are not child windows like regular Win32 programs. If you use Spy++ tool, you will know that DirectX is used to draw the UI.

**Comment by DarranRowe** [comment, marked as answer]
Yes, this is by design.
Only [Microsoft.UI.Xaml.Window](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window?view=windows-app-sdk-1.2) is documented to implement IWindowNative. Basically, only a Window is backed by a Windows API HWND, UIElements aren't, they are more than likely backed by a composition visual element.

The intended way to get graphics content into Xaml is via [SurfaceImageSource](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.surfaceimagesource?view=windows-app-sdk-1.2), [VirtualSurfaceImageSource](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.imaging.virtualsurfaceimagesource?view=windows-app-sdk-1.2) or [SwapChainPanel](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.swapchainpanel?view=windows-app-sdk-1.2).

**Reply by Monarkos** [reply]
Even when I target the window still do the same error.
Even if I create a new Window still do the same error.

**Reply by DarranRowe** [reply]
There is very little to say if you don't show what you are doing. However, this does still work in 1.4.

As an example:

```
namespace CsMeh
{
    /// <summary>
    /// An empty window that can be used on its own or navigated to within a Frame.
    /// </summary>
    public sealed partial class MainWindow : Window
    {
        public MainWindow()
        {
            this.InitializeComponent();
        }

        private void myButton_Click(object sender, RoutedEventArgs e)
        {
            var w = WinRT.Interop.WindowNative.GetWindowHandle(this);

            myButton.Content = "After GetWindowHandle";
        }
    }
}
```

![Screenshot 2024-02-01 184344](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/d96eff04-211d-418d-bc89-cb97da04ada5)

Another thing to try, that was introduced in 1.3, is obtaining the WindowId. This allows you to verify that you are attempting to obtain the handle from the correct location.

```
        private void myButton_Click(object sender, RoutedEventArgs e)
        {
            var id = this.AppWindow.Id;
            var handle = Microsoft.UI.Win32Interop.GetWindowFromWindowId(id);

            myButton.Content = "After GetWindowFromWindowId";
        }
    }
```

The [AppWindow property](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window.appwindow?view=windows-app-sdk-1.4#microsoft-ui-xaml-window-appwindow) is exposed through Microsoft.UI.Xaml.Window. So if you use AppWindow and the code doesn't compile, then the problem is your code. If the above does work but WindowNative doesn't, then you need to post a new issue since you have discovered a bug.

**Reply by Monarkos** [reply]
Hey @DarranRowe,

Thank you for the reply.

I have been able to get, but the assumption was wrong because it actually use PlataformView and not the window...

Then know I have a problem with CoreWebView2ControllerWindowReference

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/42069979/e59fcea9-929a-4be3-8095-8f21cb89650c)

`PageInfo currentPage = CurrentUser.Pages.Where(x => x.Id == pageId).FirstOrDefault();

var plataformViewWindow = currentPage.WebPage.Parent.Parent.Parent.Parent.Parent.Parent.Parent.Parent.Handler.PlatformView;
var parentWindow = WinRT.Interop.WindowNative.GetWindowHandle(plataformViewWindow);
var currentPageW2 = currentPage.WebPage.Handler.PlatformView as WebView2;
Microsoft.UI.WindowId windowId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(parentWindow);
AppWindow appWindow = Microsoft.UI.Windowing.AppWindow.GetFromWindowId(windowId);

_windowReference = CoreWebView2ControllerWindowReference.CreateFromWindowHandle(windowId.Value);
var _windowReference2 = CoreWebView2ControllerWindowReference.FromAbi(parentWindow);
`

**Comment by frankayars** [comment]
That all makes sense and appreciate the responses. I will look into the proper way to get a graphics context noted by @DarranRowe.

**Comment by Gaoyifei1011** [comment]
> We are currently encountering an exception with this call:
> 
> IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
> 
> The exception message is 'The specified cast is not valid'. The issue is not limited to the parent control. Any UIElement based object passed into that call will cause the exception. Our current project requires the underlying hWnd handle to integrate with 3rd party 3D graphics APIs and any information to resolve this exception would be appreciated.
> 
> We have tested different versions of the Windows App SDK (microsoft.windowsappsdk) and get the same result.

you could use (IntPtr)AppWindow.Id.Value
